### DISCUSSION TITLE
How to allow curly braces({,}) in request URI?

### DISCUSSION THREAD

**Discussion by mdsuhaibsa** [netty/netty#14415] [General]
Hi Team,

Netty is not allowing any request which comes to our application which has curly braces({,}) in its request URI. It throws URISyntaxException stating Illegal character in path. I need to allow these curly braces({,}) in request URI for a specific usecase. Is there a way to achieve that in Netty ? Could you please help.

[be01b293, L:/10.833.91.123:8080 - R:/10.833.91.192:1125](H2 - 17) Decoding failed: REQUEST(decodeResult: failure(java.net.URISyntaxException: Illegal character in path at index 129:http://localhost:8080/applicationinfo/labs-23101800042c7200008/app-9688851526/internal-kfl-id:\{ttpp-id-23101F:kfl-type-macro:kfl-id-510692:kfl-id-length-22}), version: HTTP/1.1)\nPOST /applicationinfo/labs-23101800042c7200008/app-9688851526/internal-kfl-id:{ttpp-id-23101F:kfl-type-macro:kfl-id-510692:kfl-id-length-22} HTTP/1.1\nx-http2-scheme: <filtered>\nhost: <filtered>\ncontent-type: <filtered>\ncontent-length: <filtered>\nx-http2-stream-id: <filtered>

Thanks.

**Comment by normanmaurer** [comment]
Can you please print the whole stack trace ?

**Reply by mdsuhaibsa** [reply]
@normanmaurer This is the full exception logs we have

"thread":"epollEventLoopGroup-2-1","loggerName":"reactor.netty.http.server.HttpServerOperations","message": "[be01b293, L:/10.833.91.123:8080 - R:/10.833.91.192:1125](H2 - 17) Decoding failed: REQUEST(decodeResult: failure(java.net.URISyntaxException: Illegal character in path at index 129:http://localhost:8080/applicationinfo/labs-23101800042c7200008/app-9688851526/internal-kfl-id:\{ttpp-id-23101F:kfl-type-macro:kfl-id-510692:kfl-id-length-22}), version: HTTP/1.1)\nPOST /applicationinfo/labs-23101800042c7200008/app-9688851526/internal-kfl-id:{ttpp-id-23101F:kfl-type-macro:kfl-id-510692:kfl-id-length-22} HTTP/1.1\nx-http2-scheme: \nhost: \ncontent-type: \ncontent-length: \nx-http2-stream-id:<filtered>","threadId":109,"threadPriority":10

**Comment by slandelle** [comment]
Curly braces are not legit characters in a URI. They should be percent encoded. See spec: https://datatracker.ietf.org/doc/html/rfc3986#section-2.1

**Reply by mdsuhaibsa** [reply]
@slandelle Curly braces are not part of reserved characters
Please refer: https://datatracker.ietf.org/doc/html/rfc3986#section-2.2

**Reply by slandelle** [reply]
@mdsuhaibsa Please check https://datatracker.ietf.org/doc/html/rfc3986#section-3.3

A `path` is composed of `segments` which are composed of `pchars` where:

```
pchar           = unreserved / pct-encoded / sub-delims / ":" / "@"
unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
```

So you can't have curly braces in a path segment, they must be percent encoded. This is also true for your backslash.

**Reply by slandelle** [reply]
@mdsuhaibsa You can also check by yourself: `new java.net.URI("http://localhost:8080/applicationinfo/labs-23101800042c7200008/app-9688851526/internal-kfl-id:%5C%7Bttpp-id-23101F:kfl-type-macro:kfl-id-510692:kfl-id-length-22%7D")` will throw as well.
